File: SmartReceipts/PDF/PDFReportTable.m
//
//  ReportTable.m
//  SmartReceipts
//
//  Created by Jaanus Siim on 10/07/15.
//  Copyright (c) 2015 Will Baumann. All rights reserved.
//

#import "PDFReportTable.h"
#import "TableHeaderRow.h"

#define MINIMUM_COLUMN_WIDTH 24.0f

@interface PDFReportTable ()

@property (nonatomic, strong) IBOutlet TableHeaderRow *headerRowPrototype;
@property (nonatomic, strong) IBOutlet TableContentRow *rowOnePrototype;
@property (nonatomic, strong) IBOutlet TableContentRow *rowTwoPrototype;
@property (nonatomic, strong) NSMutableArray *rows;
@property (nonatomic, assign) NSUInteger rowsAdded;
@property (nonatomic, assign) NSUInteger rowToStart;
@property (nonatomic, assign) BOOL tableHeaderAdded;
@property (nonatomic, assign, readwrite) BOOL hasTooManyColumnsToFitWidth;

@end

@implementation PDFReportTable

#pragma mark - Initialization

- (void)awakeFromNib {
    [super awakeFromNib];

    [self setBackgroundColor:[UIColor clearColor]];

    [self.headerRowPrototype removeFromSuperview];
    [self.rowOnePrototype removeFromSuperview];
    [self.rowTwoPrototype removeFromSuperview];

    [self setClipsToBounds:YES];

    self.rows = [NSMutableArray array];
}

#pragma mark - Public

- (void)appendValues:(NSArray *)values {
    [self.rows addObject:values];
}

- (BOOL)buildTable:(CGFloat)availableSpace {
    NSMutableArray *columnsWidth = [NSMutableArray arrayWithCapacity:self.columns.count];
    NSMutableArray *titleWidth = [NSMutableArray arrayWithCapacity:self.columns.count];
    
    for (NSUInteger column = 0; column < self.columns.count; column++) {
        [columnsWidth addObject:@([self maxWidthOfColumn:column])];
        [titleWidth addObject:@(ceilf([self.headerRowPrototype widthForValue:self.columns[column]]))];
    }

    columnsWidth = [self divideRemainingWidth:columnsWidth titleWidth:titleWidth];
    
    CGFloat yOffset = 0;
    // Don't add header on view where already added
    // Happens when we decide that not enough rows in page and push all to next page.
    // In that case original table is reused, just additional rows added
    if (!self.tableHeaderAdded) {
        yOffset = [self appendRow:self.columns columnWidths:columnsWidth usingPrototype:self.headerRowPrototype yOffset:0];
        [self setTableHeaderAdded:YES];
        [self setFrameHeightTo:yOffset];
    } else {
        yOffset = CGRectGetHeight(self.frame);
    }

    for (NSUInteger row = self.rowToStart; row < self.rows.count; row++) {
        TableContentRow *prototype;
        if (row % 2 == 0) {
            prototype = self.rowOnePrototype;
        } else {
            prototype = self.rowTwoPrototype;
        }
        
        CGFloat height = [self maxHeightForRow:self.rows[row] widths:columnsWidth prototype:prototype];
        if (yOffset + height > availableSpace) {
            [self setRowsAdded:row];
            return NO;
        }

        yOffset = [self appendRow:self.rows[row] columnWidths:columnsWidth usingPrototype:prototype yOffset:yOffset];
        [self setFrameHeightTo:yOffset];
    }
    
    return YES;
}

#pragma mark - Private methods

- (void)setFrameHeightTo:(CGFloat)height {
    CGRect myFrame = self.frame;
    myFrame.size.height = height;
    [self setFrame:myFrame];
}

- (NSMutableArray *)divideRemainingWidth:(NSMutableArray *)columns titleWidth:(NSArray *)titleWidth {
    NSNumber *width = [columns valueForKeyPath:@"@sum.self"];
    CGFloat columnsWidth = [width floatValue];
    if (columnsWidth <= CGRectGetWidth(self.frame)) {
        columns = [self divideRemainingEqually:columns columnsWidth:columnsWidth];
    } else {
        columns = [self fitColumnsWidth:columns titleWidth:titleWidth];
    }

    return columns;
}

- (NSMutableArray *)fitColumnsWidth:(NSMutableArray *)columns titleWidth:(NSArray *)titleWidths {
    NSMutableIndexSet *indexesToModify = [NSMutableIndexSet indexSet];
    NSMutableIndexSet *indexesNotToModify = [NSMutableIndexSet indexSet];
    for (NSUInteger index = 0; index < columns.count; index++) {
        if ([self canSqueezeColumn:self.columns[index]]) {
            [indexesToModify addIndex:index];
        } else {
            [indexesNotToModify addIndex:index];
        }
    }

    NSNumber *totalWidthOfNotModified = [[columns objectsAtIndexes:indexesNotToModify] valueForKeyPath:@"@sum.self"];
    NSNumber *totalWidthOfModified = [[columns objectsAtIndexes:indexesToModify] valueForKeyPath:@"@sum.self"];
    NSNumber *modifiedTitleWidths = [[titleWidths objectsAtIndexes:indexesToModify] valueForKeyPath:@"@sum.self"];
    CGFloat spaceToDivide = CGRectGetWidth(self.frame) - totalWidthOfNotModified.floatValue - modifiedTitleWidths.floatValue;

    NSMutableArray *modifiedColumns = [NSMutableArray arrayWithCapacity:columns.count];
    for (NSUInteger index = 0; index < columns.count; index++) {
        if ([indexesNotToModify containsIndex:index]) {
            [modifiedColumns addObject:columns[index]];
            continue;
        }

        NSNumber *columnWidth = columns[index];
        CGFloat percent = columnWidth.floatValue / totalWidthOfModified.floatValue;
        NSNumber *titleWidth = titleWidths[index];
        float squeezedTitleWidth = titleWidth.floatValue + floorf(spaceToDivide * percent);
        
        // check if all the columns can be placed within the table (if there enough available width)
        if (squeezedTitleWidth < MINIMUM_COLUMN_WIDTH) {
            // user able to ignore this warning and continue rendering pages if needed
            LOGGER_WARNING(@"Too many table columns to fit this page");
            self.hasTooManyColumnsToFitWidth = YES;
        }
        
        [modifiedColumns addObject:@(squeezedTitleWidth)];
    }

    return modifiedColumns;
}

- (BOOL)canSqueezeColumn:(NSString *)title {
    return [NSLocalizedString(@"receipt.column.comment", nil) isEqualToString:title]
            || [NSLocalizedString(@"receipt.column.name", nil) isEqualToString:title]
            || [NSLocalizedString(@"receipt.column.report.comment", nil) isEqualToString:title]
            || [NSLocalizedString(@"receipt.column.blank.column", nil) isEqualToString:title];
}

- (NSMutableArray *)divideRemainingEqually:(NSMutableArray *)columns columnsWidth:(CGFloat)columnsWidth {
    CGFloat remainder = CGRectGetWidth(self.frame) - columnsWidth;
    CGFloat toAdd = ceilf(remainder / (CGFloat)columns.count);
    NSMutableArray *added = [NSMutableArray arrayWithCapacity:columns.count];
    for (NSUInteger index = 0; index < columns.count; index++) {
        NSNumber *number = columns[index];
        if (toAdd > remainder) {
            toAdd = remainder;
        }

        [added addObject:@(number.floatValue + toAdd)];
        remainder -= toAdd;
    }

    return added;
}

- (CGFloat)appendRow:(NSArray *)array columnWidths:(NSArray *)widths usingPrototype:(TableContentRow *)prototype yOffset:(CGFloat)yOffset {
    CGFloat height = [self maxHeightForRow:array widths:widths prototype:prototype];
    CGFloat xOffset = 0;
    for (NSUInteger column = 0; column < array.count; column++) {
        NSString *value = array[column];
        TableContentRow *cell = (TableContentRow *) [self duplicate:prototype];
        [cell setValue:value];

        CGFloat width = [widths[column] floatValue];

        CGRect frame = cell.frame;
        frame.origin.x = xOffset;
        frame.origin.y = yOffset;
        frame.size.width = width;
        frame.size.height = height;
        [cell setFrame:frame];
        [self addSubview:cell];

        xOffset += width;
    }

    return yOffset + height;
}

- (CGFloat)maxHeightForRow:(NSArray *)array widths:(NSArray *)widths prototype:(TableContentRow *)prototype {
    CGFloat max = 0;

    for (NSUInteger column = 0; column < array.count; column++) {
        NSString *value = array[column];
        CGFloat width = [widths[column] floatValue];
        max = MAX(max, [prototype heightForValue:value usingWidth:width]);
    }

    return max;
}

- (UIView *)duplicate:(UIView *)view {
    NSData *data = [NSKeyedArchiver archivedDataWithRootObject:view];
    return [NSKeyedUnarchiver unarchiveObjectWithData:data];
}

- (CGFloat)maxWidthOfColumn:(NSInteger)column {
    CGFloat maxWidth = 0;
    NSString *title = self.columns[column];
    maxWidth = MAX(maxWidth, [self.headerRowPrototype widthForValue:title]);
    for (NSArray *row in self.rows) {
        NSString *value = row[column];
        maxWidth = MAX(maxWidth, [self.rowOnePrototype widthForValue:value]);
    }

    return ceilf(maxWidth);
}

@end

File: SmartReceipts/PDF/PDFReportTable.swift
//
//  PDFReportTable.swift
//  SmartReceipts
//
//  Created by Bogdan Evsenev on 01/12/2017.
//  Copyright © 2017 Will Baumann. All rights reserved.
//

import Foundation

fileprivate let MINIMUM_COLUMN_WIDTH: CGFloat = 24

class PDFReportTable: UIView {
    @IBOutlet var headerRowPrototype: TableHeaderRow!
    @IBOutlet var rowOnePrototype: TableContentRow!
    @IBOutlet var rowTwoPrototype: TableContentRow!
    @IBOutlet var footerRowPrototype: TableFooterRow!
    
    var title: String?
    var rows = [[String]]()
    var columns = [String]()
    var footers = [String]()
    var rowsAdded: Int = 0
    var rowToStart: Int = 0
    var tableHeaderAdded = false
    var hasTooManyColumnsToFitWidth = false
    
    override func awakeFromNib() {
        super.awakeFromNib()
        backgroundColor = UIColor.clear
        headerRowPrototype.removeFromSuperview()
        rowOnePrototype.removeFromSuperview()
        rowTwoPrototype.removeFromSuperview()
        footerRowPrototype.removeFromSuperview()
        
        clipsToBounds = true
        headerRowPrototype.contentLabel?.font = PDFFontStyle.defaultBold.font
    }
    
    func append(values: [String]) {
        rows.append(values)
    }
    
    func appendRow(row array: [String], coloumnsWidths: [CGFloat], usingPrototype: TableContentRow, yOffset: CGFloat) -> CGFloat {
        let height = maxHeightForRow(array: array, widths: coloumnsWidths, prototype: usingPrototype)
        var xOffset: CGFloat = 0
        
        for i in 0..<array.count {
            let value = array[i]
            let cell = duplicate(view: usingPrototype) as? TableContentRow
            cell?.setValue(value)
            let width = coloumnsWidths[i]
            var cellFrame = cell!.frame
            cellFrame.origin.x = xOffset
            cellFrame.origin.y = yOffset
            cellFrame.size.width = width
            cellFrame.size.height = height
            cell?.frame = cellFrame
            addSubview(cell!)
            xOffset += width
        }
        return yOffset + height
    }
    
    func setTable(title: String, yOffset: CGFloat) -> CGFloat {
        let label = UILabel(frame: CGRect.zero)
        label.frame.origin.y = yOffset
        label.font = PDFFontStyle.defaultBold.font
        label.textColor = AppTheme.reportPDFStyleColor
        label.text = title
        label.sizeToFit()
        addSubview(label)
        return yOffset + label.bounds.height
    }
    
    func buildTable(availableSpace: CGFloat) -> Bool {
        var columnsWidth = [CGFloat]()
        var titleWidth = [CGFloat]()
        
        for column in 0..<columns.count {
            columnsWidth.append(maxWidthOfColumn(column))
            let w = CGFloat(ceilf(Float(headerRowPrototype.widthFor(value: columns[column]))))
            titleWidth.append(w)
        }
        
        columnsWidth = divideRemainingWidth(columns: columnsWidth, titleWidth: titleWidth)
        
        var yOffset: CGFloat = 10
        // Don't add header on view where already added
        // Happens when we decide that not enough rows in page and push all to next page.
        // In that case original table is reused, just additional rows added
        
        if let tableTitle = title {
            yOffset = setTable(title: tableTitle, yOffset: yOffset) + 8
            title = nil
        }
        
        if !tableHeaderAdded {
            yOffset = appendRow(row: columns, coloumnsWidths: columnsWidth, usingPrototype: headerRowPrototype, yOffset: yOffset)
            tableHeaderAdded = true
            setFrameHeight(to: frame.height)
        } else {
            yOffset = frame.height
        }
        
        for row in rowToStart..<rows.count {
            var prototype: TableContentRow!
            if row % 2 == 0 {
                prototype = rowOnePrototype
            } else {
                prototype = rowTwoPrototype
            }
            
            let height = maxHeightForRow(array: rows[row], widths: columnsWidth, prototype: prototype)
            if yOffset + height > availableSpace {
                rowsAdded = row
                return false
            }
            yOffset = appendRow(row: rows[row], coloumnsWidths: columnsWidth, usingPrototype: prototype, yOffset: yOffset)
        }
        yOffset = appendRow(row: footers, coloumnsWidths: columnsWidth, usingPrototype: footerRowPrototype, yOffset: yOffset)
        setFrameHeight(to: yOffset)
        return true
    }
    
    func maxHeightForRow(array: [String], widths: [CGFloat], prototype: TableContentRow) -> CGFloat {
        var maxHeight: CGFloat = 0
        for column in 0..<array.count {
            let value = array[column]
            let width = widths[column]
            maxHeight = max(maxHeight, prototype.heightFor(value: value, width: width))
        }
        return maxHeight
    }
    
    func maxWidthOfColumn(_ column: Int) -> CGFloat {
        var maxWidth: CGFloat = 0
        let title = columns[column]
        maxWidth = max(maxWidth, headerRowPrototype.widthFor(value: title))
        for row in rows {
            let value = row[column]
            maxWidth = max(maxWidth, rowOnePrototype.widthFor(value: value))
        }
        if !footers.isEmpty {
            maxWidth = max(maxWidth, footerRowPrototype.widthFor(value: footers[column]))
        }
        return CGFloat(ceilf(Float(maxWidth)))
    }
    
    // MARK: Private
    
    private func setFrameHeight(to height: CGFloat) {
        var myFrame = frame
        myFrame.size.height = height
        frame = myFrame
    }
    
    private func duplicate(view: UIView) -> UIView {
        let data = NSKeyedArchiver.archivedData(withRootObject: view)
        return NSKeyedUnarchiver.unarchiveObject(with: data) as! UIView
    }
    
    private func divideRemainingEqually(columns: [CGFloat], columnsWidth: CGFloat) -> [CGFloat] {
        var reminder = frame.width - columnsWidth
        var toAdd = CGFloat(ceilf(Float(reminder)/Float(columns.count)))
        var added = [CGFloat]()
        for index in 0..<columns.count {
            let number = columns[index]
            if toAdd > reminder {
                toAdd = reminder
            }
            added.append(number + toAdd)
            reminder -= toAdd
        }
        return added
    }
    
    private func divideRemainingWidth(columns: [CGFloat], titleWidth: [CGFloat]) -> [CGFloat] {
        var result = columns
        let width = sum(array: columns)
        if width <= frame.width {
            result = divideRemainingEqually(columns: columns, columnsWidth: width)
        } else {
            result = fitColumnsWidth(columns: columns, titleWidths: titleWidth)
        }
        return result
    }
    
    
    private func fitColumnsWidth(columns: [CGFloat], titleWidths: [CGFloat]) -> [CGFloat] {
        let indexesToModify = NSMutableIndexSet()
        let indexesNotToModify = NSMutableIndexSet()
        for index in 0..<columns.count {
            if canSqueezeColumn(title: self.columns[index]) {
                indexesToModify.add(index)
            } else {
                indexesNotToModify.add(index)
            }
        }
        let totalWidthOfNotModified = sum(array: objectcsAt(indexes: indexesNotToModify, in: columns))
        let modifiedTitleWidths = sum(array: objectcsAt(indexes: indexesToModify, in: titleWidths))
        let sumModifiedColumns = sum(array: objectcsAt(indexes: indexesToModify, in: columns))
        let totalWidthOfModified = sumModifiedColumns == 0 ? 1 : sumModifiedColumns
        
        let spaceToDivide = frame.width - totalWidthOfNotModified - modifiedTitleWidths
        
        var modifiedColumns = [CGFloat]()
        for i in 0..<columns.count {
            if indexesNotToModify.contains(i) {
                modifiedColumns.append(columns[i])
                continue
            }
            let columnWidth = columns[i]
            let percent = columnWidth/totalWidthOfModified
            let titleWidth = titleWidths[i]
            let squeezedTitleWidth = titleWidth + CGFloat(floorf(Float(spaceToDivide * percent)))
            
            // check if all the columns can be placed within the table (if there enough available width)
            if squeezedTitleWidth < MINIMUM_COLUMN_WIDTH {
                Logger.warning("Too many table columns to fit this page")
                hasTooManyColumnsToFitWidth = true
            }
            modifiedColumns.append(squeezedTitleWidth)
        }
        
        return modifiedColumns
    }
    
    
    private func sum(array: [CGFloat]) -> CGFloat {
        var result: CGFloat = 0
        for i in 0..<array.count {
            result += array[i]
        }
        return result
    }
    
    private func objectcsAt(indexes: NSMutableIndexSet, in array: [CGFloat]) -> [CGFloat] {
        var result = [CGFloat]()
        indexes.forEach { index in
            if index < array.count {
                result.append(array[index])
            }
        }
        return result
    }
    
    private func canSqueezeColumn(title: String) -> Bool {
        return LocalizedString("receipt.column.comment") == title ||
            LocalizedString("receipt.column.name") == title ||
            LocalizedString("receipt.column.report.comment") == title ||
            LocalizedString("receipt.column.blank.column") == title
    }
    
}
